File: tools/analyze/symbolizer.cpp
// tools/analyze/symbolizer.cpp — eu-addr2line-based frame symbolizer

#include "tools/analyze/symbolizer.h"

#include <sys/stat.h>

#include <array>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <format>
#include <map>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

#include "absl/status/statusor.h"
#include "tools/analyze/log_parser.h"

namespace crashomon {
namespace {

struct FrameRef {
  uint32_t tid;
  int index;
  uint64_t offset;
};

// Run `cmd` and return its stdout. Returns empty string on error.
// The command is built by the caller; no shell metacharacters should appear
// in untrusted values (module paths come from tombstone text).
// Google C++ Style Guide recommends trailing
// return types only when required; conventional notation is clearer here.
// NOLINTNEXTLINE(modernize-use-trailing-return-type)
std::string RunCapture(const std::string& cmd) {
  // popen is intentional: this tool's purpose is to invoke addr2line
  // as a subprocess and capture its stdout.
  // NOLINTNEXTLINE(cert-env33-c, misc-include-cleaner)
  FILE* pipe = popen(cmd.c_str(), "r");  // popen/pclose via <cstdio>; include-cleaner FP.
  if (pipe == nullptr) {
    return "";
  }
  constexpr size_t read_buf_size = 256;
  std::string out;
  std::array<char, read_buf_size> buf{};
  while (fgets(buf.data(), static_cast<int>(buf.size()), pipe) != nullptr) {
    out += buf.data();
  }
  const int exit_code = pclose(pipe);  // NOLINT(misc-include-cleaner) — pclose comes via <cstdio>
                                       // which is included; false positive from include-cleaner.
  if (exit_code != 0) {
    return "";
  }
  return out;
}

// Escape a path so it is safe to embed in a shell command.
// We only allow paths that contain no shell-special characters;
// anything else is wrapped in single quotes with embedded ' escaped.
// Google C++ Style Guide recommends trailing
// return types only when required; conventional notation is clearer here.
// NOLINTNEXTLINE(modernize-use-trailing-return-type)
std::string ShellQuote(const std::string& str) {
  std::string out = "'";
  for (const char chr : str) {
    if (chr == '\'') {
      out += "'\\''";
    } else {
      out += chr;
    }
  }
  out += "'";
  return out;
}

// Google C++ Style Guide recommends trailing
// return types only when required; conventional notation is clearer here.
// NOLINTNEXTLINE(modernize-use-trailing-return-type)
bool FileExists(const std::string& path) {
  struct stat stat_buf {};
  return ::stat(path.c_str(), &stat_buf) == 0 && S_ISREG(stat_buf.st_mode);
}

// Call addr2line once for all offsets belonging to one module.
// Returns lines in order: 2 lines per address (function, then file:line).
// Google C++ Style Guide recommends trailing
// return types only when required; conventional notation is clearer here.
// NOLINTNEXTLINE(modernize-use-trailing-return-type)
std::vector<std::string> QueryAddrLine(const std::string& binary, const std::string& addr2line,
                                       const std::vector<FrameRef>& refs) {
  if (refs.empty()) {
    return {};
  }

  // Build: addr2line -e <module> -f -s -p <offset1> <offset2> ...
  // -f: print function name  -s: strip directory  -p: pretty-print (one line)
  // We use plain -f (two lines per addr) for simpler parsing.
  std::string cmd = ShellQuote(addr2line) + " -e " + ShellQuote(binary) + " -f -s";
  for (const auto& ref : refs) {
    cmd += std::format(" 0x{:016x}", ref.offset);
  }
  cmd += " 2>/dev/null";

  const std::string raw = RunCapture(cmd);
  std::vector<std::string> lines;
  std::istringstream stream(raw);
  std::string line;
  while (std::getline(stream, line)) {
    lines.push_back(line);
  }
  return lines;
}

// func_line/loc_line are distinct addr2line output fields; conventional return type notation is
// clearer per Google Style Guide.
// NOLINTNEXTLINE(bugprone-easily-swappable-parameters,modernize-use-trailing-return-type)
SymbolInfo ParseAddrLineOutput(const std::string& func_line, const std::string& loc_line) {
  SymbolInfo sym;
  sym.function = func_line;
  if (sym.function.empty()) {
    sym.function = "??";
  }

  // loc_line is "file.c:142" or "??:0"
  const auto colon = loc_line.rfind(':');
  if (colon != std::string::npos) {
    sym.source_file = loc_line.substr(0, colon);
    const std::string num_str = loc_line.substr(colon + 1);
    char* end = nullptr;
    const long val =
        std::strtol(num_str.c_str(), &end,
                    10);  // NOLINT(misc-include-cleaner) — std::strtol comes via <cstdlib> which is
                          // included; false positive from include-cleaner.
    sym.source_line = (end != num_str.c_str()) ? static_cast<int>(val) : 0;
  }
  if (sym.source_file == "??") {
    sym.source_file.clear();
  }
  return sym;
}

}  // namespace

// Google C++ Style Guide recommends trailing
// return types only when required; conventional notation is clearer here.
// NOLINTNEXTLINE(modernize-use-trailing-return-type)
absl::StatusOr<SymbolTable> SymbolizeWithAddrLine(const ParsedTombstone& tombstone,
                                                  std::string_view addr2line_binary) {
  // Group frames by module path.
  std::map<std::string, std::vector<FrameRef>> by_module;
  for (const auto& thread : tombstone.threads) {
    for (const auto& frame : thread.frames) {
      if (frame.module_path.empty()) {
        continue;
      }
      by_module[frame.module_path].push_back(
          {.tid = thread.tid, .index = frame.index, .offset = frame.module_offset});
    }
  }

  SymbolTable result;
  const std::string binary(addr2line_binary);

  for (const auto& [module, refs] : by_module) {
    if (!FileExists(module)) {
      // Binary not found on this machine — mark frames as unknown.
      for (const auto& ref : refs) {
        result[{ref.tid, ref.index}] = {"??", "", 0};
      }
      continue;
    }

    const std::vector<std::string> lines = QueryAddrLine(module, binary, refs);

    // Expect 2 lines per address.
    for (size_t i = 0; i < refs.size(); ++i) {
      const size_t base = i * 2;
      const std::string func = (base < lines.size()) ? lines[base] : "??";
      const std::string loc = (base + 1 < lines.size()) ? lines[base + 1] : "";
      result[{refs[i].tid, refs[i].index}] = ParseAddrLineOutput(func, loc);
    }
  }

  return result;
}

// function formats multiple frame types across threads; splitting would reduce clarity.
// Conventional return type notation follows Google Style Guide.
// NOLINTNEXTLINE(readability-function-cognitive-complexity,modernize-use-trailing-return-type)
std::string FormatSymbolicated(const ParsedTombstone& tombstone, const SymbolTable& symbols) {
  static constexpr std::string_view kSep =
      "*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***";

  std::ostringstream out;
  out << kSep << "\n";

  out << std::format("pid: {}, tid: {}, name: {}  >>> {} <<<\n", tombstone.pid,
                     tombstone.crashing_tid, tombstone.process_name, tombstone.process_name);

  // Signal line.
  const auto slash = tombstone.signal_info.find(" / ");
  const std::string sig_name =
      (slash != std::string::npos) ? tombstone.signal_info.substr(0, slash) : tombstone.signal_info;
  const std::string code_name =
      (slash != std::string::npos) ? tombstone.signal_info.substr(slash + 3) : "";

  if (code_name.empty()) {
    out << std::format("signal {} ({}), fault addr 0x{:016x}\n", tombstone.signal_number, sig_name,
                       tombstone.fault_addr);
  } else {
    out << std::format("signal {} ({}), code {} ({}), fault addr 0x{:016x}\n",
                       tombstone.signal_number, sig_name, tombstone.signal_code, code_name,
                       tombstone.fault_addr);
  }

  if (!tombstone.timestamp.empty()) {
    out << "timestamp: " << tombstone.timestamp << "\n";
  }

  auto emit_frames = [&](const ParsedThread& thread) {
    for (const auto& frame : thread.frames) {
      const auto sym_it = symbols.find({thread.tid, frame.index});
      const auto& mod =
          frame.module_path.empty() ? std::string_view{"???"} : std::string_view{frame.module_path};

      if (sym_it != symbols.end() && sym_it->second.function != "??") {
        const auto& sym = sym_it->second;
        if (!sym.source_file.empty() && sym.source_line > 0) {
          out << std::format("    #{:02d} pc 0x{:016x}  {} ({}) [{}:{}]\n", frame.index,
                             frame.module_offset, mod, sym.function, sym.source_file,
                             sym.source_line);
        } else {
          out << std::format("    #{:02d} pc 0x{:016x}  {} ({})\n", frame.index,
                             frame.module_offset, mod, sym.function);
        }
      } else if (frame.module_path.empty()) {
        out << std::format("    #{:02d} pc 0x{:016x}  ???\n", frame.index, frame.module_offset);
      } else if (!frame.trailing.empty()) {
        out << std::format("    #{:02d} pc 0x{:016x}  {} {}\n", frame.index, frame.module_offset,
                           frame.module_path, frame.trailing);
      } else {
        out << std::format("    #{:02d} pc 0x{:016x}  {}\n", frame.index, frame.module_offset,
                           frame.module_path);
      }
    }
  };

  // Crashing thread.
  if (!tombstone.threads.empty() && tombstone.threads[0].is_crashing) {
    out << "\nbacktrace:\n";
    emit_frames(tombstone.threads[0]);
  }

  // Other threads.
  for (size_t i = 1; i < tombstone.threads.size(); ++i) {
    const auto& thread = tombstone.threads[i];
    if (thread.name.empty()) {
      out << std::format("\n--- --- --- thread {} --- --- ---\n", thread.tid);
    } else {
      out << std::format("\n--- --- --- thread {} ({}) --- --- ---\n", thread.tid, thread.name);
    }
    emit_frames(thread);
  }

  if (!tombstone.minidump_path.empty()) {
    out << "\nminidump saved to: " << tombstone.minidump_path << "\n";
  }
  out << kSep << "\n";
  return out.str();
}

}  // namespace crashomon

File: tools/analyze/symbolizer.py
"""Symbolization and tombstone formatting utilities.

Provides:
  - parse_stackwalk_json: convert minidump-stackwalk --json output to ParsedTombstone
  - format_symbolicated: format a ParsedTombstone with symbol information
  - format_raw_tombstone: format a ParsedTombstone without symbols
  - read_minidump_annotations: read Crashpad SimpleStringDictionary from a minidump binary
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from pathlib import Path

from .log_parser import ParsedFrame, ParsedThread, ParsedTombstone

_SEP = "*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***"

# Signal name → signal number (POSIX, x86-64 Linux).
_SIGNAL_NUMBERS: dict[str, int] = {
    "SIGABRT": 6,
    "SIGBUS": 7,
    "SIGFPE": 8,
    "SIGILL": 4,
    "SIGKILL": 9,
    "SIGSEGV": 11,
    "SIGSYS": 31,
    "SIGTRAP": 5,
}


# ---------------------------------------------------------------------------
# Symbol store symbolization
# ---------------------------------------------------------------------------


@dataclass
class SymbolInfo:
    function: str = "??"
    source_file: str = ""
    source_line: int = 0


# Maps (tid, frame_index) -> SymbolInfo
SymbolTable = dict[tuple[int, int], SymbolInfo]


# ---------------------------------------------------------------------------
# rust-minidump JSON → ParsedTombstone
# ---------------------------------------------------------------------------


def parse_stackwalk_json(data: dict) -> tuple[ParsedTombstone, SymbolTable]:
    """Convert minidump-stackwalk --json output to (ParsedTombstone, SymbolTable).

    Raises ValueError if the JSON does not contain a threads array.
    """
    if "threads" not in data:
        raise ValueError("No threads array in minidump-stackwalk JSON output")

    result = ParsedTombstone()
    table: SymbolTable = {}

    result.pid = data.get("pid") or 0

    crash_info = data.get("crash_info", {})
    sig_name = crash_info.get("type") or ""
    result.signal_info = sig_name
    result.signal_number = _SIGNAL_NUMBERS.get(sig_name.split()[0], 0) if sig_name else 0
    addr_str = crash_info.get("address") or "0x0"
    try:
        result.fault_addr = int(addr_str, 16)
    except ValueError:
        result.fault_addr = 0
    crashing_thread_idx = crash_info.get("crashing_thread") or 0

    main_module_idx = data.get("main_module") or 0
    modules = data.get("modules") or []
    if main_module_idx < len(modules):
        result.process_name = Path(modules[main_module_idx].get("filename") or "").name

    for thread_idx, thread_data in enumerate(data["threads"]):
        tid = thread_data.get("thread_id") or thread_idx
        name = thread_data.get("thread_name") or ""
        thread = ParsedThread(
            tid=tid,
            name=name,
            is_crashing=(thread_idx == crashing_thread_idx),
        )
        for frame_data in thread_data.get("frames") or []:
            frame_idx = frame_data.get("frame", len(thread.frames))
            offset_str = frame_data.get("module_offset") or "0x0"
            try:
                offset = int(offset_str, 16)
            except ValueError:
                offset = 0
            frame = ParsedFrame(
                index=frame_idx,
                module_offset=offset,
                module_path=frame_data.get("module") or "",
                trust=frame_data.get("trust") or "",
            )
            thread.frames.append(frame)
            func = frame_data.get("function")
            if func:
                sym = SymbolInfo(function=func)
                src_file = frame_data.get("file")
                src_line = frame_data.get("line")
                if src_file:
                    sym.source_file = src_file
                if src_line is not None:
                    sym.source_line = int(src_line)
                table[(tid, frame_idx)] = sym
        result.threads.append(thread)

    if not result.threads:
        raise ValueError("No thread frames found in minidump-stackwalk JSON output")

    # Move crashing thread to front.
    if crashing_thread_idx < len(result.threads):
        crashing = result.threads.pop(crashing_thread_idx)
        result.threads.insert(0, crashing)
        result.crashing_tid = crashing.tid

    return result, table


# ---------------------------------------------------------------------------
# Tombstone formatter
# ---------------------------------------------------------------------------


def _emit_stack_trace_section(
    thread: ParsedThread,
    symbols: SymbolTable,
    out: list[str],
    addr_w: int,
) -> None:
    for frame in thread.frames:
        module = frame.module_path or "???"
        addr = f"0x{frame.module_offset:0{addr_w}x}"
        sym = symbols.get((thread.tid, frame.index))

        trailing = ""
        if sym and sym.function not in ("??", ""):
            if sym.source_file and sym.source_line > 0:
                trailing = f"({sym.function}) [{sym.source_file}:{sym.source_line}]"
            else:
                trailing = f"({sym.function})"
        elif frame.trailing:
            trailing = frame.trailing

        heuristic = " [HEURISTIC]" if frame.trust == "scan" else ""
        if trailing:
            out.append(f"    #{frame.index:02d} pc {addr}  {module}  {trailing}{heuristic}\n")
        else:
            out.append(f"    #{frame.index:02d} pc {addr}  {module}{heuristic}\n")


def _tombstone_addr_width(tombstone: ParsedTombstone) -> int:
    for thread in tombstone.threads:
        for frame in thread.frames:
            if frame.module_offset > 0xFFFFFFFF:
                return 16
    return 8


def format_symbolicated(tombstone: ParsedTombstone, symbols: SymbolTable) -> str:
    """Format a ParsedTombstone with resolved symbols into tombstone text."""
    addr_w = _tombstone_addr_width(tombstone)
    out: list[str] = []
    out.append(_SEP + "\n")
    out.append(
        f"pid: {tombstone.pid}, tid: {tombstone.crashing_tid},"
        f" name: {tombstone.process_name}  >>> {tombstone.process_name} <<<\n"
    )

    slash = tombstone.signal_info.find(" / ")
    if slash != -1:
        sig_name = tombstone.signal_info[:slash]
        code_name = tombstone.signal_info[slash + 3 :]
        out.append(
            f"signal {tombstone.signal_number} ({sig_name}),"
            f" code {tombstone.signal_code} ({code_name}),"
            f" fault addr 0x{tombstone.fault_addr:x}\n"
        )
    else:
        out.append(
            f"signal {tombstone.signal_number} ({tombstone.signal_info}),"
            f" fault addr 0x{tombstone.fault_addr:x}\n"
        )

    if tombstone.timestamp:
        out.append(f"timestamp: {tombstone.timestamp}\n")

    if tombstone.abort_message:
        if tombstone.terminate_type:
            out.append(f"Abort message: '{tombstone.terminate_type}: {tombstone.abort_message}'\n")
        else:
            out.append(f"Abort message: '{tombstone.abort_message}'\n")

    if tombstone.threads and tombstone.threads[0].is_crashing:
        out.append("\nbacktrace:\n")
        _emit_stack_trace_section(tombstone.threads[0], symbols, out, addr_w)

    for thread in tombstone.threads[1:]:
        if thread.name:
            out.append(f"\n--- --- --- thread {thread.tid} ({thread.name}) --- --- ---\n")
        else:
            out.append(f"\n--- --- --- thread {thread.tid} --- --- ---\n")
        _emit_stack_trace_section(thread, symbols, out, addr_w)

    if tombstone.minidump_path:
        out.append(f"\nminidump saved to: {tombstone.minidump_path}\n")

    out.append(_SEP + "\n")
    return "".join(out)


def format_raw_tombstone(tombstone: ParsedTombstone) -> str:
    """Format a ParsedTombstone without symbol information."""
    return format_symbolicated(tombstone, {})


# ---------------------------------------------------------------------------
# Crashpad simple-annotations reader
# ---------------------------------------------------------------------------

_MINIDUMP_MAGIC = b"MDMP"
_CRASHPAD_INFO_STREAM_TYPE = 0x43500007


def _read_utf8_string(data: bytes, rva: int) -> str:
    """Read a MINIDUMP_UTF8_STRING at *rva*: length(u32le) + UTF-8 bytes."""
    if rva == 0 or rva + 4 > len(data):
        return ""
    length = struct.unpack_from("<I", data, rva)[0]
    if length == 0 or rva + 4 + length > len(data):
        return ""
    return data[rva + 4 : rva + 4 + length].decode("utf-8", errors="replace")


def read_minidump_annotations(dmp_path: str) -> dict[str, str]:
    """Return the Crashpad SimpleStringDictionary from a minidump as a plain dict.

    Parses the CrashpadInfo stream (type 0x43500007) directly from the file
    binary. Returns an empty dict on any parse error or when the stream is absent.
    """
    try:
        with open(dmp_path, "rb") as fobj:
            data = fobj.read()
    except OSError:
        return {}

    if len(data) < 32 or data[:4] != _MINIDUMP_MAGIC:
        return {}

    try:
        _, _, stream_count, dir_rva = struct.unpack_from("<IIII", data, 0)

        stream_rva = 0
        for i in range(stream_count):
            off = dir_rva + i * 12
            stype, _size, srva = struct.unpack_from("<III", data, off)
            if stype == _CRASHPAD_INFO_STREAM_TYPE:
                stream_rva = srva
                break

        if stream_rva == 0:
            return {}

        # MinidumpCrashpadInfo:
        #   version(4) | report_id(16) | client_id(16) |
        #   annotations_DataSize(4) | annotations_RVA(4) | ...
        if stream_rva + 44 > len(data):
            return {}
        version = struct.unpack_from("<I", data, stream_rva)[0]
        if version != 1:
            return {}
        dict_size, dict_rva = struct.unpack_from("<II", data, stream_rva + 36)
        if dict_rva == 0 or dict_size < 4:
            return {}

        count = struct.unpack_from("<I", data, dict_rva)[0]
        result: dict[str, str] = {}
        for i in range(count):
            entry_off = dict_rva + 4 + i * 8
            if entry_off + 8 > len(data):
                break
            key_rva, val_rva = struct.unpack_from("<II", data, entry_off)
            result[_read_utf8_string(data, key_rva)] = _read_utf8_string(data, val_rva)

        return result
    except struct.error:
        return {}
